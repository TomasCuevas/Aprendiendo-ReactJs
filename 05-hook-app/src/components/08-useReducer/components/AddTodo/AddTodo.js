import React from 'react'
import PropTypes from 'prop-types';

import { useForm } from '../../../../hooks/useForm';

export const AddTodo = ({ handleAddTodo }) => {
    const [{description}, handleInputChange, reset] = useForm({ description: '' });

    const handleSubmitForm = ( e ) => {
        e.preventDefault();

        if ( description.trim().length <= 2 ) return;
        
        const newToDo = {
            id: new Date().getTime(),
            description: description,
            done: false
        }

        handleAddTodo( newToDo );
        reset();
    }

    return (
        <div className='add__todo-container'>
            <h4 className='add__todo-title'>Agregar ToDo</h4>
            <hr/>

            <form 
                autoComplete='off'
                onSubmit={ handleSubmitForm }
                className='add__todo-form'>
                <input 
                    placeholder='ToDo description'
                    className='add__todo-input'
                    type="text"
                    value={ description }
                    onChange={ handleInputChange }
                    name='description'/>
                <button 
                    type='submit'
                    className='add__todo-btn'>
                        Agregar
                </button>
            </form>
        </div>
    )
}

AddTodo.propTypes = {
    handleAddTodo: PropTypes.func.isRequired
}