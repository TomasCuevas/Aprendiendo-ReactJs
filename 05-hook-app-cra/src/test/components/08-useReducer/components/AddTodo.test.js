import { shallow } from "enzyme";
import { AddTodo } from "../../../../components/08-useReducer/components/AddTodo/AddTodo";

describe('Pruebas en <AddTodo />', () => {

  const handleAddTodo = jest.fn();
  
  const wrapper = shallow(
    <AddTodo 
      handleAddTodo={ handleAddTodo }
    />
  )
  
  test('debe de mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('no debe de llamarse la funcion handleAddTodo', () => {
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault(){} });

    expect(handleAddTodo).toHaveBeenCalledTimes(0);
  });

  test('debe de llamarse las funciones handleAddTodo y reset', () => {
    const value = 'Aprender NodeJs';
    const inputForm = wrapper.find('input');
    inputForm.simulate('change', { 
      target: {
          value: value,
          name: 'description'
        } 
      });

    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault(){} });

    expect(handleAddTodo).toHaveBeenCalledTimes(1);
    expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      description: value,
      done: false
    })
    
    expect(inputForm.prop('value')).toBe('');
  });
})