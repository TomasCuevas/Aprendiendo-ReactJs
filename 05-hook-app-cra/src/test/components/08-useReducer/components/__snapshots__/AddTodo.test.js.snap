// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddTodo /> debe de mostrarse correctamente 1`] = `
<div
  className="add__todo-container"
>
  <h4
    className="add__todo-title"
  >
    Agregar ToDo
  </h4>
  <hr />
  <form
    autoComplete="off"
    className="add__todo-form"
    onSubmit={[Function]}
  >
    <input
      className="add__todo-input"
      name="description"
      onChange={[Function]}
      placeholder="ToDo description"
      type="text"
      value=""
    />
    <button
      className="add__todo-btn"
      type="submit"
    >
      Agregar
    </button>
  </form>
</div>
`;
